### PR TITLE
fixes #306 - Detect when file:// access is not enabled

### LINKED ISSUE #306: Detect when file:// access is not enabled
We can check in code if the plugin has access to `file://` URLs and warn if not. We could implement a warning that people have to manually check the box on the `chrome://extensions` page.

see #172

https://developer.chrome.com/extensions/extension#method-isAllowedFileSchemeAccess

### PR DESCRIPTION
**Pull request by Tushkiz**
- shows devtools tab always
- warns if 'file URLs' permission is not enabled

### CODE REVIEW

**Comment by gaearon** [comment]
Do you know whether this problem exists in Firefox or not?

**Comment by Tushkiz** [comment]
It does not exist in Firefox

**Comment by Tushkiz** [comment]
This is how it will look.

devtools-warning-chrome

**Comment by Tushkiz** [comment]
amigos, any feedback on this?

File: shells/chrome/src/main.js [hunk 1]
```diff
@@ -21,9 +21,6 @@ function createPanelIfReactLoaded() {
   chrome.devtools.inspectedWindow.eval(`!!(
     Object.keys(window.__REACT_DEVTOOLS_GLOBAL_HOOK__._renderers).length || window.React
   )`, function(pageHasReact, err) {
-    if (!pageHasReact || panelCreated) {
```
Comment (suchipi) [inline shells/chrome/src/main.js]: Why do you return early here? With this change in place, it looks like the devtools will load the React tab on pages where React isn't present. This could be undesirable.

Comment (Tushkiz) [inline shells/chrome/src/main.js]: There are two cases as of now when the devtools does not display React tab
1. If React is not present
2. If React is present but the devtools extension do not have permission to access 'file://' urls

I agree that it is undesirable to show React tab for 1st case,
but for 2nd case (which behaves same as 1st, when there are no permissions) we have no clue why the tab was not displayed

That is why with this change, devtools will always display React tab.
It will show a message "**Unable to find React…**" for 1st case,
and a more specific message "**Please enable 'Allow access to file URLs' permission for the devtools extension**" for 2nd case

Comment (suchipi) [inline shells/chrome/src/main.js]: Gotcha, that makes sense now. I think this change in behavior is fine, then.

File: frontend/Panel.js [hunk 2]
```diff
@@ -310,6 +324,12 @@ var styles = {
     padding: 30,
     flex: 1,
   },
+  errorMessage: {
+    color: '#f38b00',
```
Comment (gaearon) [inline frontend/Panel.js]: I don't think we use any color in other places. Let's keep the style consistent?

File: shells/chrome/src/main.js (hunk 1 above)
Comment (gaearon) [inline shells/chrome/src/main.js]: Do we need to leave the `if (panelCreated)` check? Was it guarding against some sort of race condition?

File: frontend/Panel.js [hunk 3]
```diff
@@ -234,7 +239,16 @@ class Panel extends React.Component {
       );
     }
     if (!this.state.isReact) {
-      return <div style={styles.loading}><h2>Looking for React…</h2></div>;
+      return (
+        <div style={styles.loading}>
+          <h2 style={styles.errorMessage}>Unable to find React…</h2>
```
Comment (gaearon) [inline frontend/Panel.js]: This does not necessarily mean it’s unable to find React. As you can see it could also mean it is “looking for React” (doesn’t it always take a while to connect?) Especially with React Native.

We need to avoid showing a strongly negative message when we still might get a connection. Can you distinguish loading state from failed state somehow? Or maybe reword to be more ambiguous and imply it takes a few seconds to load if React is there.

File: shells/chrome/src/main.js (hunk 1 above)
Comment (Tushkiz) [inline shells/chrome/src/main.js]: Yes, it is guarding against [`onNavigated`](https://github.com/Tushkiz/react-devtools/blob/195ecd46e2fb1fbbf1bf7afd392f8f53c06fd2ea/shells/chrome/src/main.js#L46-L48) and [`setInterval`](https://github.com/Tushkiz/react-devtools/blob/195ecd46e2fb1fbbf1bf7afd392f8f53c06fd2ea/shells/chrome/src/main.js#L52-L54) callbacks

File: frontend/Panel.js (hunk 2 above)
Comment (Tushkiz) [inline frontend/Panel.js]: oh I see, no colors.

File: frontend/Panel.js (hunk 3 above)
Comment (Tushkiz) [inline frontend/Panel.js]: I thought the code [here](https://github.com/Tushkiz/react-devtools/blob/195ecd46e2fb1fbbf1bf7afd392f8f53c06fd2ea/frontend/Panel.js#L209-L222) could distinguish loading vs failed state, but it seems that the failed state isn't actually 'failed'.

How about we reword the messages as
- Connecting to React…
- Looking for React…
  - Please check if the devtools extension has 'Allow access to file URLs' permission

Comment (gaearon) [inline frontend/Panel.js]: Yea, it's a though one. I'll try to look at other solutions as this can potentially get more confusing.

Comment (Tushkiz) [inline frontend/Panel.js]: yeah, do let me know.

File: frontend/Panel.js [hunk 4]
```diff
@@ -234,7 +239,16 @@ class Panel extends React.Component {
       );
     }
     if (!this.state.isReact) {
-      return <div style={styles.loading}><h2>Looking for React…</h2></div>;
+      return (
+        <div style={styles.loading}>
+          <h2>Looking for React…</h2>
+          {showFileUrlWarningMessage &&
+            <div style={styles.troubleshootingHint}>
+              Please check if the devtools extension has 'Allow access to file URLs' permission
```
Comment (gaearon) [inline frontend/Panel.js]: Can we change this to [this wording](https://github.com/facebook/react-devtools/blob/60c3e1bfa3ce60a22f1de1f3d8ea2beda3df2a2e/shells/chrome/popups/not-detected.html#L6) and somehow open that link in a new tab if it's clicked?

Comment (Tushkiz) [inline frontend/Panel.js]: I have changed the message, please check.

**Comment by Tushkiz** [comment]
I completely forgot about this PR. Btw I have updated it with the latest master, @gaearon can you look into this (or suggest on the confusing bits)

**Comment by bvaughn** [comment]
React DevTools has been rewritten and [recently launched a new version 4 UI](https://reactjs.org/blog/2019/08/15/new-react-devtools.html) 🎉 The source code for this rewrite was done in a separate repository and is [being merged into the main React repo](https://github.com/facebook/react/pull/16381).

**Thank you for taking the time to contribute to this extension** 🙇‍♂️

Because this PR is for the version code base, I am closing it. If you'd like to contribute to the new extension, please visit [the new repository](https://github.com/facebook/react/issues?q=is%3Aissue+is%3Aopen+label%3A%22Component%3A+Developer+Tools%22).
